### PR TITLE
[SIW-1560] Add queue triggered function for email sender

### PR DESCRIPTION
**Pull request by christianatzeni**
**DISCLAIMER**
The code of this PR isn't ready for a production environment, because the `getUserInfoByFiscalCode()` function has a to do (it isn't implemented).

#### List of Changes

What did I do?
I created a storage queue handler that send an email to a specific user when a wallet instance has been created. This PR is based on [this PR](https://github.com/pagopa/io-wallet/pull/322).

#### Motivation and Context

The change is necessary because the user who created and activated a new wallet instance must be able to receive an email confirming it and advising that it can be revoked at any time.

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/io-wallet-user-func/src/app/main.ts [hunk 1]
```diff
@@ -236,3 +236,12 @@ app.storageQueue("validateWalletInstance", {
...
   queueName: config.azure.queue.walletInstanceRevocation.name,
 });
+
+app.storageQueue("sendEmailOnWalletInstanceCreation", {
+  connection: "StorageConnectionString",
+  handler: SendEmailOnWalletInstanceCreation({
+    emailNotificationService,
+    inputDecoder: WalletInstance,
+  }),
+  queueName: config.azure.queue.walletInstanceRevocation.name,
```
Comment (grausof) [inline apps/io-wallet-user-func/src/app/main.ts:246]:
```suggestion
  queueName: config.azure.queue.walletInstanceCreation.name,
```

File: apps/io-wallet-user-func/src/infra/handlers/send-email-on-wallet-instance-creation.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+export const sendEmailToUser: (
+  params: SendEmailNotificationParams,
+) => RTE.ReaderTaskEither<
+  { emailNotificationService: EmailNotificationService },
+  Error,
+  void
+> =
+  (params) =>
+  ({ emailNotificationService }) =>
+    pipe(params, emailNotificationService.sendEmail);
```
Comment (grausof) [inline apps/io-wallet-user-func/src/infra/handlers/send-email-on-wallet-instance-creation.ts:32]:
```suggestion
    emailNotificationService.sendEmail(params)
```

File: apps/io-wallet-user-func/src/infra/http/handlers/create-wallet-instance.ts [hunk 3]
```diff
@@ -106,23 +79,6 @@ export const CreateWalletInstanceHandler = H.of((req: H.HttpRequest) =>
...
-                      firstName,
-                      WALLET_ACTIVATION_EMAIL_FAQ_LINK,
-                      WALLET_ACTIVATION_EMAIL_HANDLE_ACCESS_LINK,
-                    ),
-                    subject: WALLET_ACTIVATION_EMAIL_SUBJECT,
-                    text: WALLET_ACTIVATION_EMAIL_TEXT,
-                    to: email,
-                  }),
-                ),
-              ),
```
Comment (grausof) [inline apps/io-wallet-user-func/src/infra/http/handlers/create-wallet-instance.ts:124]: Now I think you might want to add the Wallet Instance queue writing here too

File: apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance-creation.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import * as J from "fp-ts/lib/Json";
+import * as TE from "fp-ts/lib/TaskEither";
+import { pipe } from "fp-ts/lib/function";
+import { WalletInstance } from "io-wallet-common/wallet-instance";
+
+export interface IWalletInstanceCreationQueue {
+  insert: (walletInstance: WalletInstance) => TE.TaskEither<Error, void>;
+}
+
+export class WalletInstanceCreationStorageQueue
```
Comment (grausof) [inline apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance-creation.ts:12]: since `wallet-instance-revocation.ts` has the same code, how about we standardize them and use a single class?

Comment (christianatzeni) [reply apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance-creation.ts:12]: We could:
- delete one between the two class/interfaces (it's indifferent, the code is the same anyway!)
- rename the remaining class/interface to `WalletInstanceStorageQueue`
- use the renamed class/interface in both the revocation and creation process, obviously passing the correct data (storage queue name) during instantiation

Comment (grausof) [reply apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance-creation.ts:12]: yes you could rename it to WalletInstanceQueue and use that

File: apps/io-wallet-user-func/local.settings.json.example [hunk 5]
```diff
@@ -29,6 +29,7 @@
     "EntityConfigurationStorageAccount__serviceUri": "ENTITY_CONFIGURATION_STORAGE_ACCOUNT__SERVICE_URI",
     "EntityConfigurationStorageContainerName": "ENTITY_CONFIGURATION_STORAGE_CONTAINER_NAME",
     "StorageConnectionString": "STORAGE_ACCOUNT_CONNECTION_STRING",
+    "CreationQueueName": "WALLET_INSTANCE_CREATION_QUEUE_NAME",
```
Comment (silvicir) [inline apps/io-wallet-user-func/local.settings.json.example:32]: specify creation of the wallet instance in the queue name env

File: apps/io-wallet-user-func/src/app/config.ts [hunk 6]
```diff
@@ -278,6 +282,7 @@ export const getAzureConfigFromEnvironment: RE.ReaderEither<
   sequenceS(RE.Apply)({
     appInsights: getAzureAppInsightsConfigFromEnvironment,
     cosmos: getAzureCosmosConfigFromEnvironment,
+    creationQueueName: readFromEnvironment("CreationQueueName"),
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/app/config.ts:285]: ditto

File: apps/io-wallet-user-func/local.settings.json.example (hunk 5 above)
Comment (christianatzeni) [reply apps/io-wallet-user-func/local.settings.json.example:32]: I thought about it! Can I do the same for `RevocationQueueName`?

File: apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts [hunk 7]
```diff
@@ -24,7 +24,7 @@ export class WalletInstanceRevocationStorageQueue
         TE.tryCatch(
           () =>
             this.#client.sendMessage(walletInstanceString, {
-              visibilityTimeout: 60 * 60 * 24, //24 hours
+              visibilityTimeout: 60 * 60 * 24, // 24 hours
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts:27]: in that way we have a visibility timeout even for sending the email

File: apps/io-wallet-user-func/src/app/config.ts [hunk 8]
```diff
@@ -278,10 +282,13 @@ export const getAzureConfigFromEnvironment: RE.ReaderEither<
...
     appInsights: getAzureAppInsightsConfigFromEnvironment,
     cosmos: getAzureCosmosConfigFromEnvironment,
+    creationQueueName: readFromEnvironment("WalletInstanceCreationQueueName"),
     entityConfigurationStorageContainerName: readFromEnvironment(
       "EntityConfigurationStorageContainerName",
     ),
-    revocationQueueName: readFromEnvironment("RevocationQueueName"),
+    revocationQueueName: readFromEnvironment(
+      "WalletInstanceRevocationQueueName",
+    ),
```
Comment (grausof) [inline apps/io-wallet-user-func/src/app/config.ts:291]: Be careful here because if renamed it must also be renamed in terraform.

File: apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts
Comment (grausof) [inline apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts:1]: Sorry, I was thinking that maybe it is better not to unify these two queues. I would leave the revocation queue as it was and create a new queue for sending emails but as suggested by @silvicir I would insert in the queue only the information useful for sending the email i.e. the userId

**Review by grausof — CHANGES_REQUESTED** [review]

File: apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts
Comment (silvicir) [reply apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts:1]: maybe we can delete the `WalletInstanceStorageQueue` and define a generic `sendMessage` function
`sendMessage: (message: string) => RTE.ReaderTaskEither< { queueClient: QueueClient }, StorageQueueError, void>`
or something like that

Comment (grausof) [reply apps/io-wallet-user-func/src/infra/azure/queue/wallet-instance.ts:1]: Agree

File: apps/io-wallet-user-func/src/app/main.ts [hunk 9]
```diff
@@ -230,9 +240,18 @@ app.storageQueue("validateWalletInstance", {
...
     walletInstanceRepository,
   }),
   queueName: config.azure.queue.walletInstanceRevocation.name,
 });
+
+app.storageQueue("sendEmailOnWalletInstanceCreation", {
+  connection: "StorageConnectionString",
+  handler: SendEmailOnWalletInstanceCreationFunction({
+    emailNotificationService,
+    inputDecoder: WalletInstanceCreationEntry,
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/app/main.ts:254]: I think only `FiscalCode` is enough

File: infra/resources/_modules/function_apps/locals.tf [hunk 10]
```diff
@@ -36,7 +36,7 @@ locals {
...
       PidIssuerHealthCheckEnabled   = true
 
-      RevocationQueueName = var.revocation_queue_name
+      WalletInstanceRevocationQueueName = var.revocation_queue_name
```
Comment (grausof) [inline infra/resources/_modules/function_apps/locals.tf:39]: deletion should usually be done after deployment and switching to the new variable to avoid inconsistencies between the function deployment and the infra deployment. For this PR, both should be left and post-deployment an ad-hoc PR should be done to eliminate unused variables.

File: apps/io-wallet-user-func/local.settings.json.example [hunk 11]
```diff
@@ -29,6 +29,8 @@
     "EntityConfigurationStorageAccount__serviceUri": "ENTITY_CONFIGURATION_STORAGE_ACCOUNT__SERVICE_URI",
     "EntityConfigurationStorageContainerName": "ENTITY_CONFIGURATION_STORAGE_CONTAINER_NAME",
     "StorageConnectionString": "STORAGE_ACCOUNT_CONNECTION_STRING",
+    "WalletInstanceCreationQueueName": "WALLET_INSTANCE_CREATION_QUEUE_NAME",
```
Comment (silvicir) [inline apps/io-wallet-user-func/local.settings.json.example:32]: Don't we want to give the queue a name that refers to the fact that it is sending the email rather than referring to the creation of the wallet instance?

File: apps/io-wallet-user-func/src/app/config.ts [hunk 12]
```diff
@@ -126,6 +126,12 @@ export type AttestationServiceConfiguration = t.TypeOf<
 const AzureConfig = t.type({
   appInsights: AzureAppInsightsConfig,
   cosmos: AzureCosmosConfig,
+  queue: t.type({
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/app/config.ts:129]: this `queue` property should be removed and the properties have to be in `storage.walletInstances.queues`

File: apps/io-wallet-user-func/src/app/main.ts [hunk 13]
```diff
@@ -255,3 +260,12 @@ app.storageQueue("callPidIssuerRevokeApi", {
   queueName:
     config.azure.storage.walletInstances.queues.pidIssuerRevokeApi.name,
 });
+
+app.storageQueue("sendEmailOnWalletInstanceCreationQueue", {
+  connection: "StorageConnectionString",
+  handler: SendEmailOnWalletInstanceCreationFunction({
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/app/main.ts:266]: The handler function's name should match the one in the function app (expect for `Function`)
btw I would prefer `sendEmailOnWalletInstanceCreation`

File: apps/io-wallet-user-func/src/infra/azure/storage/queue.ts [hunk 14]
```diff
@@ -26,7 +26,7 @@ const sendMessage =
   > =>
   ({ queueClient }) =>
     pipe(
-      TE.tryCatch(() => queueClient.sendMessage(message), E.toError),
+      TE.tryCatch(async () => queueClient.sendMessage(message), E.toError),
       TE.filterOrElse(
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/infra/azure/storage/queue.ts:30]: why did you add `async`?

File: apps/io-wallet-user-func/src/infra/azure/storage/wallet-instance-revocation.ts [hunk 15]
```diff
@@ -1,13 +1,18 @@
-import { WalletInstanceRevocationQueue } from "@/wallet-instance-revocation-process";
 import { QueueClient } from "@azure/storage-queue";
 import * as E from "fp-ts/lib/Either";
 import * as J from "fp-ts/lib/Json";
 import * as TE from "fp-ts/lib/TaskEither";
 import { pipe } from "fp-ts/lib/function";
 import { WalletInstanceValidWithAndroidCertificatesChain } from "io-wallet-common/wallet-instance";
 
+export interface IWalletInstanceRevocationStorageQueue {
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/infra/azure/storage/wallet-instance-revocation.ts:8]: why did you move here the interface?

File: apps/io-wallet-user-func/src/infra/handlers/send-email-on-wallet-instance-creation.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+import WalletInstanceActivationEmailTemplate from "@/templates/wallet-instance-activation/index.html";
+import * as H from "@pagopa/handler-kit";
+import { FiscalCode } from "@pagopa/ts-commons/lib/strings";
+import { pipe } from "fp-ts/function";
+import * as RTE from "fp-ts/lib/ReaderTaskEither";
+
+import { EmailNotificationService } from "../email-notification-service";
+
+// [SIW-1560] to do - a mock function that return the user email by the fiscal code
+const getUserInfoByFiscalCode = (
```
Comment (silvicir) [inline apps/io-wallet-user-func/src/infra/handlers/send-email-on-wallet-instance-creation.ts:17]:
```suggestion
const getUserEmailByFiscalCode = (
```

File: apps/io-wallet-user-func/src/infra/azure/storage/queue.ts (hunk 14 above)
Comment (christianatzeni) [reply apps/io-wallet-user-func/src/infra/azure/storage/queue.ts]: If I don't use the async keyword, the unit test labeled "should return a HTTP 204 response on success" fails. This is probably just a vitest issue, but it seems necessary to pass all unit tests.

File: apps/io-wallet-user-func/src/infra/azure/storage/wallet-instance-revocation.ts (hunk 15 above)
Comment (christianatzeni) [reply apps/io-wallet-user-func/src/infra/azure/storage/wallet-instance-revocation.ts]: It's a typo!

File: apps/io-wallet-user-func/src/infra/azure/storage/queue.ts (hunk 14 above)
Comment (silvicir) [reply apps/io-wallet-user-func/src/infra/azure/storage/queue.ts]: I think the unit tests are wrong
